pkg/driver/kubernetes.go

```diff
@@ -0,0 +1,372 @@
+	pods                  coreclientv1.PodInterface
+	deletionPolicy        metav1.DeletionPropagation
+	requiredCompletions   int32
+}
+
+// NewKubernetesDriver initializes a Kubernetes driver.
+func NewKubernetesDriver(namespace, serviceAccount string, conf *rest.Config) (*KubernetesDriver, error) {
+	driver := &KubernetesDriver{
+		Namespace:          namespace,
+		ServiceAccountName: serviceAccount,
+	}
+	driver.setDefaults()
+	err := driver.setClient(conf)
+	return driver, err
+}
+
+// Handles receives an ImageType* and answers whether this driver supports that type.
+func (k *KubernetesDriver) Handles(imagetype string) bool {
+	return imagetype == driver.ImageTypeDocker || imagetype == driver.ImageTypeOCI
+}
+
+// Config returns the Kubernetes driver configuration options.
+func (k *KubernetesDriver) Config() map[string]string {
+	return map[string]string{
+		"KUBE_NAMESPACE":  "Kubernetes namespace in which to run the invocation image",
+		"SERVICE_ACCOUNT": "Kubernetes service account to be mounted by the invocation image (if empty, no service account token will be mounted)",
+		"KUBE_CONFIG":     "Absolute path to the kubeconfig file",
+		"MASTER_URL":      "Kubernetes master endpoint",
+	}
+}
+
+// SetConfig sets Kubernetes driver configuration.
+func (k *KubernetesDriver) SetConfig(settings map[string]string) {
+	k.setDefaults()
+	k.Namespace = settings["KUBE_NAMESPACE"]
+	k.ServiceAccountName = settings["SERVICE_ACCOUNT"]
+
+	var kubeconfig string
+	if kpath := settings["KUBE_CONFIG"]; kpath != "" {
+		kubeconfig = kpath
+	} else if home := homeDir(); home != "" {
+		kubeconfig = filepath.Join(home, ".kube", "config")
+	}
+
+	conf, err := clientcmd.BuildConfigFromFlags(settings["MASTER_URL"], kubeconfig)
+	if err != nil {
+		panic(err)
+	}
+	err = k.setClient(conf)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (k *KubernetesDriver) setDefaults() {
+	k.SkipCleanup = false
+	k.BackoffLimit = 0
+	k.ActiveDeadlineSeconds = 300
+	k.requiredCompletions = 1
+	k.deletionPolicy = metav1.DeletePropagationBackground
+}
+
+func (k *KubernetesDriver) setClient(conf *rest.Config) error {
+	coreClient, err := coreclientv1.NewForConfig(conf)
+	if err != nil {
+		return err
+	}
+	batchClient, err := batchclientv1.NewForConfig(conf)
+	if err != nil {
+		return err
+	}
+	k.jobs = batchClient.Jobs(k.Namespace)
+	k.secrets = coreClient.Secrets(k.Namespace)
+	k.pods = coreClient.Pods(k.Namespace)
+
+	return nil
+}
+
+// Run executes the operation inside of the invocation image.
+func (k *KubernetesDriver) Run(op *driver.Operation) error {
+	if k.Namespace == "" {
+		return fmt.Errorf("KUBE_NAMESPACE is required")
+	}
+	labelMap := generateLabels(op)
+	meta := metav1.ObjectMeta{
+		Namespace:    k.Namespace,
+		GenerateName: generateNameTemplate(op),
+		Labels:       labelMap,
+	}
+	// Mount SA token if a non-zero value for ServiceAccountName has been specified
+	mountServiceAccountToken := k.ServiceAccountName != ""
```

Comment:
This was an intentional decision, as we don't know that the kubernetes driver is being used to run invocation images that deploy to the same cluster, or even to kubernetes. I would like to avoid giving permissions to invocation images that haven't been explicitly declared as required by the bundle.

That said, duffle could afford to be more opinionated and set a default service account here:
https://github.com/deislabs/duffle/blob/206a2bec45ddddb9dcf558581f27249a80db835f/pkg/driver/lookup.go#L14-L15

Thoughts?

Reply:
@sbawaska also did you try mounting kubeconfig as a credential as requested by the riff bundle?

Reply:
I think the missing piece here is a lack of contract between the driver and the bundle for the parameters needed by the driver.

To me, the use case for this driver is to run the invocationImage in cluster where the bundle is to be installed. I think we should let the default docker driver handle the deploying to non-kubernetes environments. 

I like this quote from Allan Kay: 
> Simple things should be simple, complex things should be possible.

In this context it would mean: provide defaults, but allow the bundle author to override the defaults. more concretely, I think we should:
1. provide `default` as the default value for SERVICE_ACCOUNT. The bundle author could set this as a parameter and optionally provide a default value of their own in bundle.json.
2. provide default namespace from `/var/run/secrets/kubernetes.io/serviceaccount/namespace` for KUBE_NAMESPACE, again the bundle author could set this as a parameter too.

Reply:
On second though, we could iterate on these issues to make the experience better, and work on some contract between driver and duffle for parameters needed. Let's merge this!